modules/xfem/src/userobjects/MeshCut2DUserObjectBase.C

```diff
@@ -0,0 +1,334 @@
+  // _crack_front_definition->updateNumberOfCrackFrontPoints(_crack_front_points.size())
+  if (number_crack_front_points != _original_and_current_front_node_ids.size())
+    mooseError("MeshCut2DFractureUserObject::getCrackFrontPoints:  number_crack_front_points=" +
+               Moose::stringify(number_crack_front_points) +
+               " does not match the number of nodes given in "
+               "_original_and_current_front_node_ids=" +
+               Moose::stringify(_original_and_current_front_node_ids.size()));
+  for (unsigned int i = 0; i < number_crack_front_points; ++i)
+  {
+    dof_id_type id = _original_and_current_front_node_ids[i].second;
+    Node * this_node = _cutter_mesh->node_ptr(id);
+    mooseAssert(this_node, "Node is NULL");
+    Point & this_point = *this_node;
+    crack_front_points[i] = this_point;
+  }
+  return crack_front_points;
+}
+
+// CrackFrontDefinition wants the normal so this gives it the normal for a line element with an
+// assumed tangent direction in the [001] direction.
+const std::vector<RealVectorValue>
+MeshCut2DUserObjectBase::getCrackPlaneNormals(unsigned int number_crack_front_points) const
+{
+  if (number_crack_front_points != _original_and_current_front_node_ids.size())
+    mooseError("MeshCut2DFractureUserObject::getCrackPlaneNormals: number_crack_front_points=" +
+               Moose::stringify(number_crack_front_points) +
+               " does not match the number of nodes given in "
+               "_original_and_current_front_node_ids=" +
+               Moose::stringify(_original_and_current_front_node_ids.size()) +
+               ".  This will happen if a crack front exits the boundary because the number of "
+               "points in the CrackFrontDefinition is never updated.");
+
+  std::vector<std::pair<dof_id_type, RealVectorValue>> crack_plane_normals;
+  for (const auto & elem : _cutter_mesh->element_ptr_range())
+  {
+    dof_id_type id0 = elem->node_id(0);
+    dof_id_type id1 = elem->node_id(1);
+    dof_id_type id;
+
+    auto it0 = std::find_if(_original_and_current_front_node_ids.begin(),
+                            _original_and_current_front_node_ids.end(),
+                            [&id0](const std::pair<dof_id_type, dof_id_type> & element)
+                            { return element.second == id0; });
+    auto it1 = std::find_if(_original_and_current_front_node_ids.begin(),
+                            _original_and_current_front_node_ids.end(),
+                            [&id1](const std::pair<dof_id_type, dof_id_type> & element)
+                            { return element.second == id1; });
+
+    bool found_it0 = (it0 != _original_and_current_front_node_ids.end());
+    bool found_it1 = (it1 != _original_and_current_front_node_ids.end());
+    if (found_it0 || found_it1)
+    {
+
+      Point end_pt, connecting_pt;
+      // make sure normal crossed with the crack face points in the [0,0,1] direction
```

Comment:
This comment isn't really describing what's going on below, is it?

Reply:
It is but there is a better comment 10 lines down.